src/listener/tcp_listener.rs

```diff
@@ -0,0 +1,116 @@
+use super::is_transient_error;
+
+use crate::listener::Listener;
+use crate::utils::BoxFuture;
+use crate::{log, Server};
+
+use std::fmt::{self, Display, Formatter};
+
+use async_std::net::{self, SocketAddr, TcpStream};
+use async_std::prelude::*;
+use async_std::{io, task};
+
+#[derive(Debug)]
+pub enum TcpListener {
+    FromListener(net::TcpListener),
+    FromAddrs(Vec<SocketAddr>, Option<net::TcpListener>),
+}
```

Comment:
Hmm, I guess I'm somewhat confused by this structure. In the stdlib `ToSocketAddr` is resolved the moment it is created. Is there a practical reason we're not doing the same here?

That way we can store a single `async_std::net::TcpListener` that we then later listen on. The same goes for `unix_listener.rs`. `ToListener::to_listener` is fallible, so I don't believe it should be a problem.

Reply:
The difference is that ToSocketAddr doesn't immediately _bind_ to that address. In this version of the code, we do as much resolution as we can, but we avoid binding until listen is called. That way this doesn't exhibit confusing behavior:
```rust
let mut multi = MultiListener::new();
multi.add("localhost:8000")?;
multi.add("unix+http://file")?;
// are 8000 and ./file bound here? I wouldn't expect them to be yet

multi.listen().await?; //but now I would expect them to be.

Reply:
I also would not expect resources by string to be opened until `listen` is called, unless someone passes in something that is already explicitly bound, i.e. a struct for a file descriptor or something.